# Тур 1: задачи A, B, C, D

Разбор задач из файла [«Тур 1.pdf»](Тур%201.pdf). Все вычисления сделаны целыми числами и рациональными дробями: внешние библиотеки не требуются.

| Задача | Ответ |
|---|---|
| A | семь возможных значений; печатаются ниже в порядке возрастания |
| B | 0 70 220 |
| C | 0.235000 |
| D | 2.428571 |

## A. Натуральные числа

Пусть стёртое первое число равно \(x\). Первые члены ряда:

\[
x,\quad 1,\quad x+1,\quad 2(x+1),\quad 4(x+1),\ldots
\]

Начиная с третьего члена, каждый новый член равен сумме всех предыдущих. После его добавления эта сумма удваивается. Поэтому для \(n\ge3\)

\[
a_n=2^{n-3}(x+1).
\]

Заданное число \(N=123456123456123456\) не может стоять на втором месте, поскольку там находится 1. Разложение \(N=2^6\cdot1929001929001929\), где второй множитель нечётный. Следовательно, \(N=2^k(x+1)\) возможно только при \(k=0,1,\ldots,6\). Для каждого \(k\) получаем \(x=N/2^k-1\). Соответствующие места \(n=k+3\) лежат между 3 и 9, то есть входят в ряд из 100 чисел.

In [1]:
N = 123456123456123456
odd_part = N
power_of_two = 0
while odd_part % 2 == 0:
    odd_part //= 2
    power_of_two += 1

answers_a = sorted(
    N // (2 ** k) - 1
    for k in range(min(power_of_two, 97) + 1)
)
assert power_of_two == 6 and odd_part == 1929001929001929
assert all(x > 0 for x in answers_a)
print("A:", *answers_a)

A: 1929001929001928 3858003858003857 7716007716007715 15432015432015431 30864030864030863 61728061728061727 123456123456123455


## B. Матрицы и забытые активации

Сколько бы матриц ни было, их последовательное умножение без нелинейностей и прибавления констант задаёт **одно линейное отображение** \(T:\mathbb R^2\to\mathbb R^3\).

Все координаты данных цветов строго между 0 и 255. Значит, операция clip в этих двух точках ничего не изменила:

\[
T(-2,3)=(60,40,100),\qquad T(1,3)=(200,50,80).
\]

Векторы \((-2,3)\) и \((1,3)\) линейно независимы. Целевую точку выражаем через них:

\[
(-7,6)=3(-2,3)-(1,3).
\]

По линейности

\[
T(-7,6)=3(60,40,100)-(200,50,80)=(-20,70,220).
\]

Последний шаг — обрезать каждую координату в диапазон \([0,255]\): \(f(-7,6)=(0,70,220)\). Ответ единственный, потому что два независимых входных вектора задают линейное отображение на всей плоскости.

In [2]:
p = (-2, 3)
q = (1, 3)
target = (-7, 6)
assert tuple(3 * p[i] - q[i] for i in range(2)) == target

Tp = (60, 40, 100)
Tq = (200, 50, 80)
before_clip = tuple(3 * Tp[i] - Tq[i] for i in range(3))
answer_b = tuple(min(255, max(0, value)) for value in before_clip)
print("До clip:", before_clip)
print("B:", *answer_b)

До clip: (-20, 70, 220)
B: 0 70 220


## C. Среднее и медиана: перевод условия

Обозначим отсортированные числа через \(t_1<\dots<t_{10}\). Условие о расстоянии даёт \(t_{i+1}-t_i\ge0.01\).

Любой отрезок длины \(0.25\) содержит хотя бы одну точку тогда и только тогда, когда

\[
t_1\le0.25,\qquad t_{10}\ge0.75,\qquad
t_{i+1}-t_i\le0.25\quad(i=1,\dots,9).
\]

Иначе слева, справа или между соседними точками поместился бы пустой отрезок длины \(0.25\). Эти неравенства, наоборот, исключают такой пустой отрезок.

### Верхняя граница

Достаточно сначала оценить \(\overline t-m\). Из минимального расстояния \(0.01\):

\[
t_1+t_2+t_3+t_4\le4t_5-(0.04+0.03+0.02+0.01)=4t_5-0.10.
\]

Из максимального промежутка \(0.25\):

\[
t_7+t_8+t_9\le(t_6+0.25)+(t_6+0.50)+(t_6+0.75)
=3t_6+1.50.
\]

Также \(t_{10}\le1\). Поэтому

\[
\sum_{i=1}^{10}t_i\le5t_5+4t_6+2.40,
\]

\[
\overline t-m
\le\frac{5t_5+4t_6+2.40}{10}-\frac{t_5+t_6}{2}
=0.24-0.1t_6.
\]

Так как \(t_1\ge0\) и первые шесть точек отделены промежутками не меньше \(0.01\), имеем \(t_6\ge0.05\). Следовательно, \(\overline t-m\le0.235\). Замена каждого числа на \(1-t_i\) сохраняет условия и меняет знак разности, поэтому \(|\overline t-m|\le0.235\).

Граница достигается на наборе

\[
0,\ 0.01,\ 0.02,\ 0.03,\ 0.04,\ 0.05,\ 0.30,\ 0.55,\ 0.80,\ 1.
\]

Здесь среднее равно \(0.28\), медиана — \(0.045\), разность — \(0.235\).

In [3]:
from fractions import Fraction as F

points_c = [F(i, 100) for i in range(6)] + [
    F(30, 100), F(55, 100), F(80, 100), F(1)
]
gaps_c = [b - a for a, b in zip(points_c, points_c[1:])]
assert len(points_c) == 10
assert all(F(1, 100) <= gap <= F(1, 4) for gap in gaps_c)
assert points_c[0] <= F(1, 4) and points_c[-1] >= F(3, 4)

mean_c = sum(points_c) / 10
median_c = (points_c[4] + points_c[5]) / 2
answer_c = abs(mean_c - median_c)
assert answer_c == F(47, 200)
print("Среднее:", mean_c, "Медиана:", median_c)
print("C:", f"{float(answer_c):.6f}")

Среднее: 7/25 Медиана: 9/200
C: 0.235000


## D. MAE

Считываем 14 точек с рисунка и группируем их по \(x\):

| \(x\) | Значения \(y\) |
|---:|---|
| 0 | 0, 2, 7, 8 |
| 4 | 4, 7 |
| 6 | 1, 5 |
| 9 | 0, 1, 4, 7 |
| 11 | 2, 6 |

Для фиксированного \(x\) предсказание \(a x+b\) — одно число \(z\). Сумма \(\sum|z-y_i|\) минимальна, если \(z\) лежит **между двумя средними значениями** группы. Это свойство медианы.

| \(x\) | Интервал оптимальных предсказаний | Минимальная сумма ошибок |
|---:|---:|---:|
| 0 | [2, 7] | 13 |
| 4 | [4, 7] | 3 |
| 6 | [1, 5] | 4 |
| 9 | [1, 4] | 10 |
| 11 | [2, 6] | 4 |

Сумма этих отдельных нижних границ равна \(34\). Число **4 входит во все пять интервалов**, значит постоянная прямая \(\hat y(x)=4\) (то есть \(a=0,\ b=4\)) достигает нижней границы сразу для всех групп. Поэтому

\[
\min MAE=\frac{34}{14}=\frac{17}{7}\approx2.428571.
\]

In [4]:
groups_d = {
    0: [0, 2, 7, 8],
    4: [4, 7],
    6: [1, 5],
    9: [0, 1, 4, 7],
    11: [2, 6],
}
assert sum(len(ys) for ys in groups_d.values()) == 14

lower_bound = 0
for x, values in groups_d.items():
    ys = sorted(values)
    left_median, right_median = ys[len(ys)//2 - 1], ys[len(ys)//2]
    assert left_median <= 4 <= right_median
    group_min = sum(abs(left_median - y) for y in ys)
    achieved_by_line = sum(abs(4 - y) for y in ys)
    assert group_min == achieved_by_line
    lower_bound += group_min
    print(f"x={x:2}: интервал [{left_median}, {right_median}], сумма {group_min}")

answer_d = F(lower_bound, 14)
assert lower_bound == 34 and answer_d == F(17, 7)
print("D:", f"{float(answer_d):.6f}")

x= 0: интервал [2, 7], сумма 13
x= 4: интервал [4, 7], сумма 3
x= 6: интервал [1, 5], сумма 4
x= 9: интервал [1, 4], сумма 10
x=11: интервал [2, 6], сумма 4
D: 2.428571


## Готовые ответы

- **A:** 1929001929001928 3858003858003857 7716007716007715 15432015432015431 30864030864030863 61728061728061727 123456123456123455
- **B:** 0 70 220
- **C:** 0.235000
- **D:** 2.428571

В A нужно вывести все семь значений через пробел. В C и D сохранены требуемые шесть знаков после точки.